### DISCUSSION TITLE
New "aosmap" format

### DISCUSSION THREAD

**Discussion by NotAFile** [piqueserver/piqueserver#644] [Feature Requests]
Currently, maps are several files with no standard metadata, scripts, thumbnails, dependencies, agreed compression or anything. I think it would be worthwhile to add a new format that is a bundle of these in a zip file.

**Comment by feikname** [comment]
Do you have a straw man proposal on how we could do it?

Mine is the follwing:

```text
map_name.zip
     - map_name.vxl
     - metadata_info.txt
     - notes.txt       (optional)
     - media/          (optional)
```

Where:
 * `notes.txt` is a plain text file that may or may not be included and has general notes about the map.
 * `thumbnails`/ is an optional folder that may contain map thumbnails and has no strict naming conventions for the files. We should probably encourage for only a few formats to be used, such as JPEG, PNG, webp, GIF, SVG and BPG. Explicit discouragement for proprietary formats such as PSD should also be made.
 * `map_name.vxl` is the name of the file
 * `metadata.txt` is a txt file with the format stated below.


`metadata_info.txt` sample: 

```text
aosmap_metadata_file_version: 1 # (always an integer, no semver)
map_title: My Great Name I Came Up With
#map_title_alias_1: Another Name #optional
map_author_1: feikname 
map_author_2: random_user
recommended_game_mode_1_name: zombie_ctf
recommended_game_mode_1_script: ctf # no version specified
gm_1_depends_upon_1: zombie_on_death # no version specified
gm_1_depends_upon_2: only_two_spawn_points[v2.x.y]
gm_1_recommends_1: powerup[v3.5.x] @READ_THE_NOTES  # Warn the user to read the notes (e.g. 3.5.1 is broken)
```
if map name or authors are unknown:
```text
@UNKNOWN_MAP_NAME
@UNKNOWN_AUTHORS
```

if something is missing:
```text
missing_info: @MISSING
```

A few notes:
 * Comments begin with an `#` and go until the end of the line.
 * Trailing space is to be ignored.
 * The map file should be assumed as being the only vxl file at the root of the zip.
 * The EOL of the `txt` files should be CRLF.

I chose `txt` format because most map makers are on Windows, I assume. So it would be easy for then to edit it by hand with `notepad.exe`.

This is only an idea for us to get started.

To be honest, I believe **maps should be script agnostic.** So the format I suggested above is not one I entirely agree with.

Also, scripts may change often, but maps are usually published only once. Keep that in mind.

Perhaps something based on the INI format would be nice. I'm against using JSON or YAML as the metadats file format.

**Comment by feikname** [comment]
Also, can you please expand on the no standard compression? I'm not aware of the vxl files internals.

You mean like, some are gziped inside, some are decompressed, etc?

**Comment by xtreme8000** [comment]
What I've found to yield the best compression results is a xz-archive, that is _LZMA2_.

**Comment by NotAFile** [comment]
The issue is that most common tools don't seem to support lzma and we don't want to make making these bundles harder than necessary. Meanwhile, Zip file support ships with all OSs and also with Python. It might not be the best compression wise, but I think it is pretty hard to beat usability wise.

**Comment by damienflament** [comment]
@feikname : Why not use JSON file format ? It is already used for the server configuration file.

**Comment by NotAFile** [comment]
It is, but last time we agreed we wanted to move away from that a while ago. It's a good data interchange format, but a poor config format. No comments, weird syntax and quotes, commas, brackets. TOML or INI is a lot easier to understand for casual users.

**Comment by feikname** [comment]
@damienflament JSON is a bit of a pain to manually edit when compared to other options such as INI or TOML.  Also, the stuff notafile said above.

Regarding YAML, I've gone against it because it has some non obvious idiosyncrasies such as not allowing tabs for indentation.

**Comment by NotAFile** [comment]
yes, I disagree with YAML too. I had considered it for the config system rewrite because it is JSON compatible, so existing configs would keep working, but if you don't know the syntax it can be hard to figure out.

**Comment by xtreme8000** [comment]
_or we just stick to deflate again_

**Comment by NotAFile** [comment]
reading up, python supports lzma, bzip2 and zlib for zip file compression. This means we can have our cake and eat it too!

**Comment by feikname** [comment]
Also, an optional fog key might be nice to add, so mapmakers can select their map fog.

**Comment by NotAFile** [comment]
I think most tbings currently in the mapdata.txt (which is a python file :rolleyes:) should be moved there. That includes fog and plugins to load and stuff.

**Comment by JohnDowson** [comment]
I think it would make sense to have an archive+ uncompressed metadata appended after it. It would make it possible to use metadata to address a map by it's name instead of relying on filenames, and searching maps by gamemode or something like that.

**Comment by NotAFile** [comment]
Theoretically, yes. But that would require custom tools to create new and read existing maps. Being able to create and read files using standard tools is very important to me.

**Comment by JohnDowson** [comment]
At least with gzip it doesn't. I tested, gzip just ignores the whatever is appended after the compressed data. And it is perfectly readable, you just need to scroll past the binary data.

**Comment by JohnDowson** [comment]
So what actually need to be done, regardles of the format, is switching from redefining get_entity_location and get_spawn_location, to just using coordinates from the metadata.

**Comment by NotAFile** [comment]
gzip can read it, sure. But how is the average mapmaker, who is not very technical and uses windows, going to create a binary metadata file and append it to the end of a gzipped vxl stream.

Similarly, how is a mapper going to extract the gzipped vxl to modify it, without losing any of the metadata.

For spawn locations, I'd look at issue #509 which is a generic approach to defining locations in the map metadata.

**Comment by JohnDowson** [comment]
It's pretty simple simple. Winrar unpacks such file just fine, and metadata can be extracted with notepad, though notepad++ works better. 
Creating aosmap would be even simpler:
```
cmd.exe> type map.vxl.gz metadata.txt > map.aosmap
```
Nevermind the fact that talking about accessibility of mapping with how awkward and unfriendly map editor is, is pretty pointless.
Even coding up a GUI tool to add, edit and extract metadata would be pretty easy, easier than working in voxlap editor I'd say, the hardest part is dealing with a mess of GUI boilerplate.
And if we're talking new editor, then accessibility of the format is redundant, all the tools for working with it can be a part of the editor.

**Comment by NotAFile** [comment]
> Even coding up a GUI tool to add, edit and extract metadata would be pretty easy

It would, but as always, it's unlikely anbody is going to do it.

Yeah that command works, but everyone already knows how to open and edit archives using winrar or 7zip. I also don't really see the benefit. You can still easily open files from archives in memory, even if it might be a bit slower. And you also now need to figure out when the gzip data ends. If you then add in the ability to add map preview images like in iceball, and you're just recreating an archive format.

**Comment by JohnDowson** [comment]
> > Even coding up a GUI tool to add, edit and extract metadata would be pretty easy
> 
> It would, but as always, it's unlikely anbody is going to do it.

https://github.com/JohnDowson/aosmap-metastitcher
Here it is. Plain text+lzma
